test/addressBook.test.ts

```diff
@@ -0,0 +1,305 @@
+  })
+
+  describe('Set controller', () => {
+    let controller: UpgradeableContractV1Instance
+
+    before(async () => {
+      controller = await UpgradeableContractV1.new()
+    })
+
+    it('should revert adding otoken implementation address from non-owner address', async () => {
+      await expectRevert(
+        addressBook.setController(controller.address, {from: random}),
+        'Ownable: caller is not the owner',
+      )
+    })
+
+    it('should set controller address', async () => {
+      await addressBook.setController(controller.address, {from: owner})
+
+      const proxy: OwnedUpgradeabilityProxyInstance = await OwnedUpgradeabilityProxy.at(
+        await addressBook.getController(),
+      )
+      const implementationAddress = await proxy.implementation()
+
+      assert.equal(controller.address, implementationAddress, 'Controller implementation address mismatch')
+
+      controller = await UpgradeableContractV1.at(await addressBook.getController())
+
+      assert.equal((await controller.CONTRACT_REVISION()).toString(), '1', 'Controller implementation version mismatch')
+    })
+  })
+
+  describe('Set otoken factory', () => {
+    let otokenFactory: OtokenFactoryInstance
+
+    before(async () => {
+      otokenFactory = await OtokenFactory.new(addressBook.address)
+    })
+
+    it('should revert adding otoken factory address from non-owner address', async () => {
+      await expectRevert(
+        addressBook.setOtokenFactory(otokenFactory.address, {from: random}),
+        'Ownable: caller is not the owner',
+      )
+    })
+
+    it('should set otoken factory address', async () => {
+      await addressBook.setOtokenFactory(otokenFactory.address, {from: owner})
+
+      assert.equal(await addressBook.getOtokenFactory(), otokenFactory.address, 'Otoken factory address mismatch')
+    })
+  })
+
+  describe('Set whitelist', () => {
+    let whitelist: WhitelistInstance
+
+    before(async () => {
+      whitelist = await Whitelist.new(addressBook.address)
+    })
+
+    it('should revert adding whitelist address from non-owner address', async () => {
+      await expectRevert(
+        addressBook.setWhitelist(whitelist.address, {from: random}),
+        'Ownable: caller is not the owner',
+      )
+    })
+
+    it('should set whitelist address', async () => {
+      await addressBook.setWhitelist(whitelist.address, {from: owner})
+
+      assert.equal(await addressBook.getWhitelist(), whitelist.address, 'Whitelist address mismatch')
+    })
+  })
+
+  describe('Set margin pool', () => {
+    it('should revert adding pool address from non-owner address', async () => {
+      await expectRevert(addressBook.setMarginPool(marginPoolAdd, {from: random}), 'Ownable: caller is not the owner')
+    })
+
+    it('should set pool address', async () => {
+      await addressBook.setMarginPool(marginPoolAdd, {from: owner})
+
+      assert.equal(await addressBook.getMarginPool(), marginPoolAdd, 'Pool address mismatch')
+    })
```

Comment:
why is this not testing with a newly deployed marginPool instance?

Reply:
MarginPool contract in this PR is not implemented yet, so I used an address, it behave the same as any address as the MarginPool is not upgradeable contract.